File: kotlin/space-age/src/main/kotlin/SpaceAge.kt
import java.math.BigDecimal

class SpaceAge(val ageInSeconds: Long) {
  // Given a planet's orbital period in Earth years, returns this object's age
  // in years on that planet, rounded to two decimal places.
  private fun ageOnPlanet(orbitalPeriod: Double): Double {
    val earthYear = 31557600 // seconds
    val aYearOnThisPlanet = earthYear * orbitalPeriod
    return roundToTwoDecimalPlaces(ageInSeconds / aYearOnThisPlanet)
  }

  private fun roundToTwoDecimalPlaces(d: Double): Double {
    return BigDecimal(d).setScale(2, BigDecimal.ROUND_HALF_UP).toDouble()
  }

  fun onEarth():   Double { return ageOnPlanet(1.0) }
  fun onMercury(): Double { return ageOnPlanet(0.2408467) }
  fun onVenus():   Double { return ageOnPlanet(0.61519726) }
  fun onMars():    Double { return ageOnPlanet(1.8808158) }
  fun onJupiter(): Double { return ageOnPlanet(11.862615) }
  fun onSaturn():  Double { return ageOnPlanet(29.447498) }
  fun onUranus():  Double { return ageOnPlanet(84.016846) }
  fun onNeptune(): Double { return ageOnPlanet(164.79132) }
}

File: scala/space-age/src/main/scala/SpaceAge.scala
object SpaceAge {
  val EarthYear = 31557600 // seconds

  // Given a planet's orbital period, returns a function that expresses an age
  // in seconds in terms of years on that planet, rounded to two decimal places.
  def ageOnPlanet(orbitalPeriod: Double): (Double) => Double = {
    val yearOnPlanet = EarthYear * orbitalPeriod
    (ageInSeconds: Double) => roundToTwoPlaces(ageInSeconds / yearOnPlanet)
  }

  val onEarth   = ageOnPlanet(1.0)
  val onMercury = ageOnPlanet(0.2408467)
  val onVenus   = ageOnPlanet(0.61519726)
  val onMars    = ageOnPlanet(1.8808158)
  val onJupiter = ageOnPlanet(11.862615)
  val onSaturn  = ageOnPlanet(29.447498)
  val onUranus  = ageOnPlanet(84.016846)
  val onNeptune = ageOnPlanet(164.79132)

  private[this] def roundToTwoPlaces(d: Double): Double = {
    BigDecimal(d).setScale(2, BigDecimal.RoundingMode.HALF_UP).toDouble
  }
}
